tasks/skylab/PairedTagUtils.wdl

```diff
@@ -1,66 +1,128 @@
+        description: "Checks read2 FASTQ length and orientation and performs trimming."
     }
-
     parameter_meta {
         read1_fastq: "read 1 FASTQ files of paired reads -- forward reads"
         read3_fastq: "read 3 FASTQ files of paired reads -- reverse reads"
         barcodes_fastq: "read 2 FASTQ files which contains the cellular barcodes"
+        preindex: "Boolean for whether data has a sample barcode that needs to be demultiplexed"
+        whitelist: "Atac whitelist for 10x multiome data"
+        input_id: "Input ID to demarcate sample"
         docker: "(optional) the docker image containing the runtime environment for this task"
         mem_size: "(optional) the amount of memory (MiB) to provision for this task"
         cpu: "(optional) the number of cpus to provision for this task"
         disk_size: "(optional) the amount of disk space (GiB) to provision for this task"
-        preemptible: "(optional) if non-zero, request a pre-emptible instance and allow for this number of preemptions before running the task on a non preemptible machine"
+        preemptible: "(optional) if non-zero, request a pre-emptible instance and allow for this number of preemptions before running the task on a non preemptible machine"        
     }
-
     command <<<
-
         set -e
-        echo ~{read1_fastq}
-        echo ~{barcodes_fastq}
-        echo ~{read3_fastq}
-        echo Renaming files
+        ## Need to gunzip the r1_fastq
+        pass="true"
+        zcat ~{barcodes_fastq} | head -n2 > r2.fastq
+        FASTQ=r2.fastq
+        echo 'this is the fastq:' $FASTQ
+        R2=$(awk 'NR==2' $FASTQ)
+        COUNT=$(echo ${#R2})
+        echo 'this is the read:' $R2
+        echo 'this is the barcode count:' $COUNT
+        echo "Renaming files for UPS tools"
         mv ~{read1_fastq} "~{input_id}_R1.fq.gz"
         mv ~{barcodes_fastq} "~{input_id}_R2.fq.gz"
         mv ~{read3_fastq} "~{input_id}_R3.fq.gz"
-
-        echo Running UPStools
-        upstools sepType_DPT ~{input_id} 3
+        echo performing read2 length and orientation checks 
+        if [[ $COUNT == 27 && ~{preindex} == "false" ]]
+          then
+          echo "Preindex is false and length is 27 bp"
+          echo "Trimming first 3 bp with UPStools"
+          upstools trimfq ~{input_id}_R2.fq.gz 4 26
+          echo "Running orientation check"
+          file="~{input_id}_R2_trim.fq.gz"
+          zcat "$file" | sed -n '2~4p' | shuf -n 1000 > downsample.fq
+          head -n 1 downsample.fq
+          python3 /upstools/pyscripts/dynamic-barcode-orientation.py downsample.fq ~{whitelist} best_match.txt
+          cat best_match.txt
+          barcode_choice=$(<best_match.txt)
+          echo "Barcode choice is: "
+          echo $barcode_choice
+          if [[ $barcode_choice == "FIRST_BP_RC" ]]; then
+            echo "Correct barcode orientation"
+          else
```

Comment:
Do we need an override for this, like a boolean "ignore_barcode_orientation" or something? Maybe not needed yet? (same question for the duplicate instance below)

Reply:
This is probably a silly question, but should the barcode orientation check be after the trimming? What happens if it's not in the correct orientation? Would we trim some of the barcode?

Reply:
The fastq must be 24 bp to check for the correct 10x ATAC orientation, because it can be one of 4 options: first 16 bp, first 16 bp reverse complement, last 16 bp, and last 16 bp reverse complement, assuming a 24 bp read.

Reply:
I see. will it always choose the best of the 4 or is there a scenario in which it gets such low concordance that it fails completely? (like if we trimmed 3bp or barcode and left 3bp of preindex)

Reply:
No, it will fail if there's an unreasonably low number of reads. We based that number on the real scenarios we saw when we had the wrong barcode orientation.

Reply:
Sounds good!